### ISSUE TITLE
Padding does not work when padding tensor on device [Bug Report]

### ISSUE THREAD

**Issue by athul-bos-semi** [tenstorrent/tt-metal#7044]
When trying to pad (using `tt_lib.tensor.pad`) a tensor of input shape [1,1,n,1] to a new tensor of shape [1,1,n,2] (using 3 for this example), the output tensor is correctly padded with 3s where 3s are supposed to be, but the final output is wrong.

```
ttnn.Tensor([[[[-0.35938],
               [-0.57422],
               [-0.24609],
               [ 0.06030],
               [-0.08105],
               [-0.81641],
               [ 1.15625],
               [ 0.90625],
               [ 1.47656],
               [ 0.10352]]]], shape=Shape([1, 1, 10, 1]), dtype=DataType::BFLOAT16, layout=Layout::ROW_MAJOR)
StorageType.DEVICE

------ Padding ------

ttnn.Tensor([[[[-0.35938,  3.00000],
               [ 0.20020,  3.00000],
               [ 1.35938,  3.00000],
               [ 0.54297,  3.00000],
               [-0.58594,  3.00000],
               [ 0.85938,  3.00000],
               [ 0.17480,  3.00000],
               [ 0.89844,  3.00000],
               [-0.35938,  3.00000],
               [ 0.20020,  3.00000]]]], shape=Shape([1, 1, 10, 2]), dtype=DataType::BFLOAT16, layout=Layout::ROW_MAJOR)
StorageType.DEVICE
```

**To Reproduce**
```
import ttnn
import tt_lib

import torch

ttnn.set_printoptions(profile="full")

device_id = 3
device = tt_lib.device.CreateDevice(device_id)

activations = torch.randn((1, 1, 1, 10), dtype=torch.bfloat16)
act = tt_lib.tensor.Tensor(activations)
act = act.to(device)
act = act.reshape(1,1,act.volume(),1)

# print(dir(act))
print(act)
print(act.storage_type())
print("------ Padding ------")
act = tt_lib.tensor.pad(act, [1, 1, act.volume(), 2], [0,0,0,0], 3)
# act = tt_lib.tensor.pad(act, [1, 1, 1, act.volume()+2], [0,0,0,0], 3)
print(act)
print(act.storage_type())

tt_lib.device.CloseDevice(device)
```

**Output screenshot**
![image](https://github.com/tenstorrent-metal/tt-metal/assets/147588938/5d5d1d48-7b91-470e-99bc-1036412a56f6)

**Please complete the following environment information:**
- OS: Devcloud running on Ubuntu 20.04.4 LTS (Focal)
- Version of software: v0.46.0-rc7

**Additional context**
When padding [1,1,1,n] to [1,1,1,n+2], the output is correct.

**Comment by davorchap** [comment]
FYI @mbahnasTT

**Comment by jliangTT** [comment]
@tarafdarTT. Is this a spurious behavior? is there a low-hanging fruit solution?

**Comment by mbahnasTT** [comment]
hi @tarafdarTT , I tried to use ttnn.pad,  `act = ttnn.pad(act, ((0, 0), (0,0), (0, 0), (0, 1)), 3)` 
but it's falling back. 
What shall we do about the tt_lib OP?


![Image](https://github.com/tenstorrent/tt-metal/assets/116673264/fd2a631f-cf23-4a2c-a2f6-ed9bce50ad85)

**Comment by tarafdarTT** [comment]
@mbahnasTT looks like ttnn is falling back because of the PCC error which is probably related to this bug. Sorry I was attending to other P1s and P0s. If this is a P1 I can look at this next @jliangTT . I should have my two reshard issues sorted out by tonight and can look at this tomorrow (I think previously this was a P2). 
I don't think this is low-hanging fruit solution, but I'm not sure as I didn't originally implement this op. This will take some time to first investigate.

**Comment by jliangTT** [comment]
no problem. Yes it has been promoted to P1 just today. you didn't miss anything.

**Comment by tarafdarTT** [comment]
I figured it out. The issue isn't in the padding op (so TTNN needs to be modified to not fallback, I can do that). 

The real issue is that we need page-widths to be aligned to a word at the minimum and this was not caught because of the reshape being used is not the proper reshape, I'll post an example of which reshape we should use below. A word being 4 bytes.
The width of a page in a row-major tensor is the last dimension so for bfloat16 we need an even number for the last dim. 
if you wanted that shape then you an also change the dtype to a float32 (which I think we have support for ) or uint32 (which I know we have support for)

The following works, note that I changed the reshape. The reshape you used would work for this shape as well, but the reshape I used will properly assert out if the output shape is incorrect (1,1,10,1, dtype=bfloat16) . 

```
    activations = torch.randn((1, 1, 2, 10), dtype=torch.bfloat16)
    act = ttl.tensor.Tensor(activations)
    act = act.to(device)
    print(act)
    print(act.storage_type())
    print("------ Reshape ------")
    act = ttl.tensor.reshape(act, 1, 1, 10, 2)
    print(act)
    print(act.storage_type())

    print("------ Padding ------")
    # pad (input tensor, output shape, input start, pad value)
    act = ttl.tensor.pad(act, [1, 1, 10, 4], [0,0,0,0], 3)
    print(act)
    print(act.storage_type())
```


This is the output if you run the above code: 

![image](https://github.com/tenstorrent/tt-metal/assets/135640067/d3df4598-fddf-4086-91f8-d225b3a160dd)



TLDR:
1. Either change the shape such that the last dim after the reshape is even OR change the dtype to a 32-bit type.
2. Use the above reshape since it would have asserted with the improper shape.
3. At the ttnn level use the experimental pad `(ttnn.experimental.tensor.pad)` with the same args in my example for now since TTNN incorrectly fallsback. 

Since the posted example actually works with the correct shapes or dtypes, I'm closing this issue. 
I have made a separate to stop TTNN from falling back as the kernel actually works. 
https://github.com/tenstorrent/tt-metal/issues/8094
